fe/fe-core/src/main/java/com/starrocks/scheduler/history/TaskRunHistory.java

```diff
@@ -0,0 +1,219 @@
+    private static final Logger LOG = LogManager.getLogger(TaskRunHistory.class);
+
+    // Thread-Safe history map:
+    // QueryId -> TaskRunStatus
+    // The same task-id may contain multi history task run status, so use query_id instead.
+    private final Map<String, TaskRunStatus> historyTaskRunMap =
+            Collections.synchronizedMap(Maps.newLinkedHashMap());
+    // TaskName -> TaskRunStatus
+    private final Map<String, TaskRunStatus> taskName2Status =
+            Collections.synchronizedMap(Maps.newLinkedHashMap());
+
+    private final TaskRunHistoryTable historyTable = new TaskRunHistoryTable();
+
+    public void addHistory(TaskRunStatus status) {
+        historyTaskRunMap.put(status.getQueryId(), status);
+        taskName2Status.put(status.getTaskName(), status);
+    }
+
+    public TaskRunStatus getTaskByName(String taskName) {
+        return taskName2Status.get(taskName);
+    }
+
+    public TaskRunStatus getTask(String queryId) {
+        if (queryId == null) {
+            return null;
+        }
+        return historyTaskRunMap.get(queryId);
+    }
+
+    public void removeTaskByQueryId(String queryId) {
+        if (queryId == null) {
+            return;
+        }
+        TaskRunStatus task = historyTaskRunMap.remove(queryId);
+        taskName2Status.remove(task.getTaskName());
+    }
+
+    // Reserve historyTaskRunMap values to keep the last insert at the first.
+    public List<TaskRunStatus> getInMemoryHistory() {
+        List<TaskRunStatus> historyRunStatus = new ArrayList<>(historyTaskRunMap.values());
+        Collections.reverse(historyRunStatus);
+        return historyRunStatus;
+    }
+
+    public List<TaskRunStatus> lookupHistoryByTaskNames(String dbName, Set<String> taskNames) {
+        List<TaskRunStatus> result = getInMemoryHistory().stream()
+                .filter(x -> x.matchByTaskName(dbName, taskNames))
+                .collect(Collectors.toList());
+        if (isEnableArchiveHistory()) {
+            result.addAll(historyTable.lookupByTaskNames(dbName, taskNames));
+        }
+        return result;
+    }
+
+    public List<TaskRunStatus> lookupHistory(TGetTasksParams params) {
+        List<TaskRunStatus> result = getInMemoryHistory().stream()
+                .filter(x -> x.match(params))
+                .collect(Collectors.toList());
+        if (isEnableArchiveHistory()) {
+            result.addAll(historyTable.lookup(params));
+        }
+        return result;
+    }
+
+    // TODO: make it thread safe
+    /**
+     * Remove expired task runs, would be called in regular daemon thread, and also in checkpoint
+     */
+    public void vacuum() {
+        removeExpiredRuns();
+
+        // trigger to force gc to avoid too many history task runs.
+        forceGC();
+
+        // archive histories
+        if (!GlobalStateMgr.isCheckpointThread()) {
+            archiveHistory();
+        }
+    }
+
+    public void replay(ArchiveTaskRunsLog log) {
```

Comment:
is it better to add a try-catch since it will restart fe fail if replay log meets an exception?

Reply:
I've marked it as `IgnorableOnReplayFailed`